### ISSUE TITLE
$order->getCustomer() returns NULL for registered customer

### ISSUE THREAD

**Issue by jakwinkler** [magento/magento2#25268]
### Preconditions (*)

1. Magento 2.3.2 & 2.4-develop, community edition

### Steps to reproduce (*)

1. Create customer account and make an order
2. Create a module with an observer to `sales_order_save_after` event
```
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Event/etc/events.xsd">
    <event name="sales_order_save_after">
        <observer name="my-observer" instance="Something\Darkside\Observer\OrderSaveAfter" />
    </event>
</config>
```
3. Try to access Customer object from order object

```
public function execute(Observer $observer): void
    {
         /** @var Order $order */
         $order = $observer->getEvent()->getOrder();
         $customer = $order->getCustomer();
        ....
```

### Expected result (*)

1. Method `getCustomer` should return customer object, as defined in phpdoc block **OR** phpdoc block should be changed to highlight that `NULL` could be returned

### Actual result (*)

1. Method `getCustomer` always returns `NULL`, but it's not highlighted in phpdoc block that it could return null

### Additional info
Method `getCustomer` will return customer only in case if someone added it to Order object, in other cases it doesn't exists. In case if you need to get customer from the Order object you need to do following get customer ID from the order object
```php
$customerId = $order->getCustomerId();

// $customerId could be NULL in case if order was placed as guest
$customer = $customerId ? $this->customerRepository->getById($customerId) : null;
```
Where `$this->customerRepository` is instance of `\Magento\Customer\Api\CustomerRepositoryInterface` that retrieved via constructor

**Comment by jakwinkler** [comment]
Any update on this?

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @itsPranith 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-22188` were created

**Issue Available**: @itsPranith, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by mohith-ecommerce** [comment]
@qsolutions-pl in this event we are not getting customer object so instead of $order->getCustomer() use like $order->getCustomerFirstname() for customer first name

**Comment by jakwinkler** [comment]
honestly @mohith227 that is not an answer nor a solution, I need customer object to fetch other customer data like:
- customer custom attributes
- customer group

**Comment by ihor-sviziev** [comment]
Hi @qsolutions-pl @itsPranith @rajeebkumar-sahoo,

Unfortunately adding new method to existing functionality will break backward compatibility, more details: https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/

Why do you think order model should have method "getCustomer"?

As for me it shoudn't as it's not his responsibility. It should just contain getCustomerId method or something like this, and it is:
https://github.com/magento/magento2/blob/bff0d385cfc6138b5a75423498bf9e9a88061d79/app/code/Magento/Sales/Model/Order.php#L2744-L2747

If you need to have customer - you can get it from customer repository.

I would mark it as not an issue

**Comment by rajeebkumar-sahoo** [comment]
Hi @ihor-sviziev,

The call to "$order->getCustomer()" as mentioned in the issue does not raise any PHP error saying that "method does not exist on object" but it returns NULL. So the method already exists and the functionality provided on "Order" model class as a magic method like 
"@method \Magento\Customer\Model\Customer getCustomer()". And the magic method gets resolved in "\Magento\Framework\DataObject::__call" where it doesn't find any key named "customer" and hence returns NULL which violets the method signature defined like "@method \Magento\Customer\Model\Customer getCustomer()".

The solution to this could be either we provide logic on "__call" by overriding it in "\Magento\Sales\Model\Order" or the way I did it by providing a direct function. I choose the second method because overriding "__call" would be only possible by introducing static text comparison like "if ($method === 'getCustomer')", which is not appropriate.

**Comment by ihor-sviziev** [comment]
In this case would be better to update phpdoc block that will show that
return type for getCustomer() could be also null. Is it sounds reasonable
to you?

**Comment by rajeebkumar-sahoo** [comment]
HI @ihor-sviziev @qsolutions-pl @itsPranith,

Yes, that sounds reasonable and also we should mark the magic method as deprecated so that with the next minor release we could remove that altogether.  

like: 
`* @method \Magento\Customer\Model\Customer|null getCustomer() { @deprecated  instead use getCustomerId() along with customer repository to get customer instance }
`

**Comment by ihor-sviziev** [comment]
Replacing to something like will be good enough: `* @method \Magento\Customer\Model\Customer|null getCustomer()`
So people will see that it could be also null and will check it additionally. Also as it's not real method - we can't deprecate it.

**Comment by rajeebkumar-sahoo** [comment]
Actually we can see "https://github.com/phpDocumentor/phpDocumentor/issues/1604" if we could or will go with the above solution.

**Comment by ihor-sviziev** [comment]
> Actually we can see "[phpDocumentor/phpDocumentor#1604](https://github.com/phpDocumentor/phpDocumentor/issues/1604)" if we could or will go with the above solution.

It wasn't fixed and AFAIK major IDEs are not supporting this syntax.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Hotel 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-25044` were created

**Issue Available**: @engcom-Hotel, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by ihor-sviziev** [comment]
@engcom-Hotel i think we expected result should be changed, that magic method should highlight that it could return null. This is not direct responsibility of order object to retrieve customer. In client code customer  could be retrieved by customerId.
Btw both methods getCustomer and getCustomerId might return null in case when order were placed as guest, and that would be absolutely valid behavior

**Comment by engcom-Hotel** [comment]
@ihor-sviziev thanks for contributing! I have updated the description.

**Comment by slavvka** [comment]
Hi @qsolutions-pl. Thank you for your report. 
The issue has been fixed in magento/magento2#26423 by @sertlab in 2.4-develop branch
Related commit(s):
 - [92816ab04db61666741ae9ab0986175068c46bee](https://github.com/magento/magento2/commit/92816ab04db61666741ae9ab0986175068c46bee)
 - [427a4ad298210ed5287044cd49c51391daa0af65](https://github.com/magento/magento2/commit/427a4ad298210ed5287044cd49c51391daa0af65)
 - [f8bcd47e0959bf34b1f2f01712c0488e0939e881](https://github.com/magento/magento2/commit/f8bcd47e0959bf34b1f2f01712c0488e0939e881)
 - [80c737a7287439876f73ab90c7fce7e1a5a43877](https://github.com/magento/magento2/commit/80c737a7287439876f73ab90c7fce7e1a5a43877)
 - [5a1f48e1cc590a9605c0ae82844e89a9b74e84f6](https://github.com/magento/magento2/commit/5a1f48e1cc590a9605c0ae82844e89a9b74e84f6)
 - [57e2532e05b93315c610ffc9076b422030c94781](https://github.com/magento/magento2/commit/57e2532e05b93315c610ffc9076b422030c94781)
 - [3ca45d8efa0a2a79c207451428b845d34c1addc8](https://github.com/magento/magento2/commit/3ca45d8efa0a2a79c207451428b845d34c1addc8)

The fix will be available with the upcoming 2.4.0 release.
